### PR TITLE
[blocked] Marquez-airflow openlineage integration

### PR DESCRIPTION
**Pull request by mobuchowski**

### CODE REVIEW

File: integrations/airflow/marquez_airflow/dag.py [hunk 1]
```diff
@@ -77,44 +74,47 @@ def create_dagrun(self, *args, **kwargs):
 
         return dagrun
 
-    def _register_dagrun(self, dagrun, execution_date, run_args):
+    # TODO: do we actually need to start tasks here or just register a hierarchy?
+    # Doing it other way would require to hook up to
+    # scheduler, where tasks are actually started
```
Comment (wslulciuc) [inline integrations/airflow/marquez_airflow/dag.py:79]: We make the assumption that when a DAG run is created, it's tasks can be _safely_ marked as **started** as well. This will, as you note, help build the graph hierarchy. But, not all tasks start at the same time. So ideally, we'd only start the task right before it's scheduled for execution by a worker.

File: integrations/airflow/marquez_airflow/extractors/__init__.py [hunk 2]
```diff
@@ -166,12 +176,12 @@ def validate(self):
         assert (self.operator_class is not None and
                 self.operator.__class__ == self.operator_class)
 
-    # TODO: Only return a single StepMetadata object.
     @abstractmethod
-    def extract(self) -> [StepMetadata]:
+    def extract(self) -> Union[StepMetadata, List[StepMetadata]]:
+        # Strongly preferred to return single StepMetadata instance
```
Comment (wslulciuc) [inline integrations/airflow/marquez_airflow/extractors/__init__.py:181]: In future releases, we'll want to deprecate returning a list of `StepMetadata` and simply return a `StepMetadata` object.

File: integrations/airflow/openlineage/client.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+
+class OpenLineageClient:
+    def __init__(
+            self,
+            url: str,
+            options: OpenLineageClientOptions = OpenLineageClientOptions(),
+            session: Session = None
+    ):
+        self.url = url  # TODO: url validation?
```
Comment (wslulciuc) [inline integrations/airflow/openlineage/client.py:27]: Would we just validate that the set value for `url` is a valid URL?

File: integrations/airflow/openlineage/constants.py [hunk 4]
```diff
@@ -10,21 +10,8 @@
...
-    def set(self, key, value):
-        self.mapping[key] = value
-
-    def pop(self, key, _session):
-        return self.mapping.pop(key)
+DEFAULT_TIMEOUT_MS = 10000
+DEFAULT_NAMESPACE_NAME = 'default'
+DEFAULT_MARQUEZ_BACKEND = 'http'
+DEFAULT_MARQUEZ_URL = 'http://localhost:5000'
+DEFAULT_PRODUCER = "marquez-airflow"  # TODO: mechanism for setting producer
```
Comment (wslulciuc) [inline integrations/airflow/openlineage/constants.py:17]: _Minor:_ We may want to namespace the environment variables for OpenLineage. For example:

```
DEFAULT_OPENLINEAGE_PRODUCER
```

File: integrations/airflow/openlineage/client.py (hunk 3 above)
Comment (mobuchowski) [reply integrations/airflow/openlineage/client.py]: Done using `urllib.urlparse`

File: integrations/airflow/marquez_airflow/extractors/__init__.py (hunk 2 above)
Comment (mobuchowski) [reply integrations/airflow/marquez_airflow/extractors/__init__.py]: Nice wording, changed :+1:

File: integrations/airflow/marquez_airflow/dag.py (hunk 1 above)
Comment (mobuchowski) [reply integrations/airflow/marquez_airflow/dag.py:80]: Agreed. Although, sounds good for a future task.

File: integrations/airflow/openlineage/client.py (hunk 3 above)
Comment (wslulciuc) [reply integrations/airflow/openlineage/client.py]: It's not a check I'd normally make, but as long as we silently handle, then log the exception I think this is ok to add.

**Comment by julienledem** [comment]
the error in the integration is strange: 
```
marquez_1            | ERROR [2021-03-23 13:01:01,890] io.dropwizard.jersey.errors.LoggingExceptionMapper: Error handling a request: df057bae06025df4
marquez_1            | ! java.lang.IllegalArgumentException: No enum constant marquez.common.models.FieldType.varchar
marquez_1            | ! at java.base/java.lang.Enum.valueOf(Unknown Source)
marquez_1            | ! at marquez.common.models.FieldType.valueOf(FieldType.java:17)
marquez_1            | ! at marquez.service.mappers.Mapper.toField(Mapper.java:305)
marquez_1            | ! at marquez.service.DatasetService.toField(DatasetService.java:394)
marquez_1            | ! at java.base/java.util.stream.ReferencePipeline$3$1.accept(Unknown Source)
marquez_1            | ! at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(Unknown Source)
marquez_1            | ! at java.base/java.util.stream.AbstractPipeline.copyInto(Unknown Source)
marquez_1            | ! at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(Unknown Source)
marquez_1            | ! at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(Unknown Source)
marquez_1            | ! at java.base/java.util.stream.AbstractPipeline.evaluate(Unknown Source)
marquez_1            | ! at java.base/java.util.stream.ReferencePipeline.collect(Unknown Source)
marquez_1            | ! at marquez.service.DatasetService.toDataset(DatasetService.java:337)
marquez_1            | ! at marquez.service.DatasetService.toDataset(DatasetService.java:316)
marquez_1            | ! at java.base/java.util.Optional.map(Unknown Source)
marquez_1            | ! at marquez.service.DatasetService.get(DatasetService.java:261)
marquez_1            | ! at marquez.api.DatasetResource.get(DatasetResource.java:109)
marquez_1            | ! at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
marquez_1            | ! at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
marquez_1            | ! at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
marquez_1            | ! at java.base/java.lang.reflect.Method.invoke(Unknown Source)
marquez_1            | ! at org.glassfish.jersey.server.model.internal.ResourceMethodInvocationHandlerFactory.lambda$static$0(ResourceMethodInvocationHandlerFactory.java:52)
marquez_1            | ! at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher$1.run(AbstractJavaResourceMethodDispatcher.java:124)
marquez_1            | ! at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.invoke(AbstractJavaResourceMethodDispatcher.java:167)
marquez_1            | ! at org.glassfish.jersey.server.model.internal.JavaResourceMethodDispatcherProvider$ResponseOutInvoker.doDispatch(JavaResourceMethodDispatcherProvider.java:176)
marquez_1            | ! at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.dispatch(AbstractJavaResourceMethodDispatcher.java:79)
marquez_1            | ! at org.glassfish.jersey.server.model.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:469)
marquez_1            | ! at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:391)
marquez_1            | ! at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:80)
marquez_1            | ! at org.glassfish.jersey.server.ServerRuntime$1.run(ServerRuntime.java:255)
marquez_1            | ! at org.glassfish.jersey.internal.Errors$1.call(Errors.java:248)
marquez_1            | ! at org.glassfish.jersey.internal.Errors$1.call(Errors.java:244)
marquez_1            | ! at org.glassfish.jersey.internal.Errors.process(Errors.java:292)
marquez_1            | ! at org.glassfish.jersey.internal.Errors.process(Errors.java:274)
marquez_1            | ! at org.glassfish.jersey.internal.Errors.process(Errors.java:244)
marquez_1            | ! at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:265)
marquez_1            | ! at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:234)
marquez_1            | ! at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:680)
marquez_1            | ! at org.glassfish.jersey.servlet.WebComponent.serviceImpl(WebComponent.java:394)
marquez_1            | ! at org.glassfish.jersey.servlet.WebComponent.service(WebComponent.java:346)
marquez_1            | ! at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:366)
marquez_1            | ! at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:319)
marquez_1            | ! at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:205)
marquez_1            | ! at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:763)
marquez_1            | ! at org.eclipse.jetty.servlet.ServletHandler$ChainEnd.doFilter(ServletHandler.java:1633)
marquez_1            | ! at io.dropwizard.servlets.ThreadNameFilter.doFilter(ThreadNameFilter.java:35)
marquez_1            | ! at org.eclipse.jetty.servlet.FilterHolder.doFilter(FilterHolder.java:193)
marquez_1            | ! at org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter(ServletHandler.java:1609)
marquez_1            | ! at io.dropwizard.jersey.filter.AllowedMethodsFilter.handle(AllowedMethodsFilter.java:47)
marquez_1            | ! at io.dropwizard.jersey.filter.AllowedMethodsFilter.doFilter(AllowedMethodsFilter.java:41)
marquez_1            | ! at org.eclipse.jetty.servlet.FilterHolder.doFilter(FilterHolder.java:193)
marquez_1            | ! at org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter(ServletHandler.java:1609)
marquez_1            | ! at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:561)
marquez_1            | ! at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:233)
marquez_1            | ! at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1434)
marquez_1            | ! at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:188)
marquez_1            | ! at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:501)
marquez_1            | ! at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:186)
marquez_1            | ! at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1349)
marquez_1            | ! at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
marquez_1            | ! at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
marquez_1            | ! at com.codahale.metrics.jetty9.InstrumentedHandler.handle(InstrumentedHandler.java:313)
marquez_1            | ! at io.dropwizard.jetty.RoutingHandler.handle(RoutingHandler.java:52)
marquez_1            | ! at org.eclipse.jetty.server.handler.gzip.GzipHandler.handle(GzipHandler.java:766)
marquez_1            | ! at org.eclipse.jetty.server.handler.RequestLogHandler.handle(RequestLogHandler.java:54)
marquez_1            | ! at org.eclipse.jetty.server.handler.StatisticsHandler.handle(StatisticsHandler.java:179)
marquez_1            | ! at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
marquez_1            | ! at org.eclipse.jetty.server.Server.handle(Server.java:516)
marquez_1            | ! at org.eclipse.jetty.server.HttpChannel.lambda$handle$1(HttpChannel.java:383)
marquez_1            | ! at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:556)
marquez_1            | ! at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:375)
marquez_1            | ! at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:273)
marquez_1            | ! at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311)
marquez_1            | ! at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105)
marquez_1            | ! at org.eclipse.jetty.io.ChannelEndPoint$1.run(ChannelEndPoint.java:104)
marquez_1            | ! at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:336)
marquez_1            | ! at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:313)
marquez_1            | ! at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:171)
marquez_1            | ! at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:129)
marquez_1            | ! at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:375)
marquez_1            | ! at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:773)
marquez_1            | ! at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:905)
marquez_1            | ! at java.base/java.lang.Thread.run(Unknown Source)
```
@wslulciuc  are those types still an enum, I thought this became just a string?

**Comment by mobuchowski** [comment]
@julienledem the integration test uses released image of marques 0.12.2, while this was fixed in https://github.com/MarquezProject/marquez/pull/1105

**Comment by wslulciuc** [comment]
> Probably next step in improving integration tests would be using marquez image build in previous build step, instead of released one

Yeah, building the marquezAPI from source to be used in the integration tests is really the right approach.

File: integrations/airflow/tests/integration/integration.py [hunk 5]
```diff
@@ -203,13 +212,15 @@ def check_jobs_meta():
...
     }]
     assert insert_job['location'] is None
-    assert insert_job['context']['sql'] is not None
-    assert insert_job['context']['airflow.operator'] == \
-           'airflow.operators.postgres_operator.PostgresOperator'
-    assert insert_job['context']['airflow.task_info'] is not None
-    assert insert_job['context']['airflow.version'] == AIRFLOW_VERSION
-    assert insert_job['context']['marquez_airflow.version'] == MARQUEZ_AIRFLOW_VERSION
-    assert insert_job['description'] == DAG_DESCRIPTION
+    # TODO: waiting for backend
```
Comment (wslulciuc) [inline integrations/airflow/tests/integration/integration.py:215]: Is this related to a Marquez backend fix?

File: integrations/airflow/tests/integration/integration.py [hunk 6]
```diff
@@ -203,13 +212,15 @@ def check_jobs_meta():
...
+    # assert insert_job['context']['sql'] is not None
+    # assert insert_job['description'] == DAG_DESCRIPTION
+    # TODO: no airflow context job data yet
```
Comment (wslulciuc) [inline integrations/airflow/tests/integration/integration.py:218]: Do we need to define airflow specific facets?

**Review by wslulciuc — APPROVED** [review]
LGTM 💯 🥇

File: integrations/airflow/tests/integration/integration.py (hunk 6 above)
Comment (mobuchowski) [reply integrations/airflow/tests/integration/integration.py:218]: I'll create PR with this tomorrow.
